x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plugin/TransportRetryAction.java

```diff
@@ -6,36 +6,57 @@
  */
 package org.elasticsearch.xpack.ql.plugin;
 
+import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
 import org.elasticsearch.action.search.SearchPhaseExecutionException;
 import org.elasticsearch.action.search.VersionMismatchException;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
 import org.elasticsearch.xpack.ql.util.Holder;
 
-import java.util.function.Consumer;
+import static org.elasticsearch.action.ActionListener.wrap;
 
-public final class TransportActionUtils {
+/**
+ * A base class for operations that will be retried in case first attempt failed with a {@code VersionMismatchException}.
+ * The second attempt will be performed on a node of an older version.
+ */
+public final class TransportRetryAction<Request extends ActionRequest, Response extends ActionResponse> {
```

Comment:
I'm a bit unsure about the name of the class. Why is it called `Action`? And it seems to perform a very specific retry. I would just call it `RetryOnVersionMismatch`.